### PR TITLE
Replica deletion service

### PR DESCRIPTION
**Pull request by bryanlb**
Deletes expired, un-assigned replicas on a configurable schedule.

### CODE REVIEW

**Review by vthacker — APPROVED** [review]
LGTM! This looks fairly similar to the other scheduled services.

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaDeletionService.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Deletes expired and unassigned replicas. If a replica is still assigned, marked for eviction, or
+ * in the process of evicting this will not delete the replica. Any replicas that are not deleted in
+ * a given run will be attempted in the following execution.
+ */
+public class ReplicaDeletionService extends AbstractScheduledService {
```
Comment (mansu) [inline kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaDeletionService.java:32]: Is there a reason to keep replica deletion and eviction a separate service? Isn't it more efficient to keep them in the same service?

**Review by mansu — CHANGES_REQUESTED** [review]

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaDeletionService.java (hunk 1 above)
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaDeletionService.java:32]: As discussed in more depth [on the eviction service pr](https://github.com/slackhq/kaldb/pull/179#discussion_r779724751) we're keeping these separate due to duration it takes to evict, and for better fault tolerance.

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaDeletionService.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+                  ListenableFuture<?> future = replicaMetadataStore.delete(replicaMetadata);
+                  addCallback(
+                      future,
+                      FutureUtils.successCountingCallback(successCounter),
+                      MoreExecutors.directExecutor());
+                  return future;
+                })
+            .collect(Collectors.toUnmodifiableList());
+
+    ListenableFuture<?> futureList = Futures.successfulAsList(replicaDeletions);
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaDeletionService.java:125]: Agree this needs refactored as [per the eviction service PR](https://github.com/slackhq/kaldb/pull/179#discussion_r779846687). Followup PR will address all services at once and move this functionality to a more general implementation.

File: config/config.yaml [hunk 3]
```diff
@@ -65,6 +65,8 @@ managerConfig:
     schedulePeriodMins: ${KALDB_MANAGER_CACHE_SLOT_PERIOD_MINS:-15}
   replicaEvictionServiceConfig:
     schedulePeriodMins: ${KALDB_MANAGER_REPLICA_EVICT_PERIOD_MINS:-15}
+  replicaDeletionServiceConfig:
+    schedulePeriodMins: ${KALDB_MANAGER_REPLICA_DELETE_PERIOD_MINS:-360}
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply config/config.yaml]: Not in particular - there's not a huge rush to delete replicas so if we need to we can bump this up. I set this back to 15 mins for now to match the eviction time.

File: kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaDeletionService.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+    Set<String> replicaIdsWithAssignments =
+        cacheSlotMetadataStore
+            .getCached()
+            .stream()
+            .map(cacheSlotMetadata -> cacheSlotMetadata.replicaId)
+            .collect(Collectors.toUnmodifiableSet());
+
+    AtomicInteger successCounter = new AtomicInteger(0);
+    long milliNow = Instant.now().toEpochMilli();
```
[the remark this answers is not in the document]
Comment (bryanlb) [reply kaldb/src/main/java/com/slack/kaldb/clusterManager/ReplicaDeletionService.java]: Done.

**Review by mansu — APPROVED** [review]

**Pull request merged**